File: generated_code/python/solution_2418_3.py
from typing import List

class Solution:
    def minSumSquareDiff(self, nums1: List[int], nums2: List[int], k1: int, k2: int) -> int:
        """
        Algorithm Explanation:
        1. The absolute differences between nums1[i] and nums2[i] are what matter. 
           Modifying an element in nums1 or nums2 is equivalent to reducing the absolute 
           difference |nums1[i] - nums2[i]| by 1. 
        2. We have a total budget of k = k1 + k2 operations to reduce these absolute differences.
        3. To minimize the sum of squares, we should greedily reduce the largest differences first 
           because squaring penalizes larger numbers much more harshly.
        4. Since the maximum possible difference is 10^5, we can use a bucket sort / frequency array 
           approach instead of a max-heap to achieve an O(N + MAX_DIFF) time complexity.
        """
        # Calculate the total number of operations we can perform
        k = k1 + k2
        
        # Max possible difference is bounded by the constraints (0 <= nums1[i], nums2[i] <= 10^5)
        max_diff = 100000
        # frequency[d] will store the number of pairs with an absolute difference of d
        frequency = [0] * (max_diff + 1)
        
        total_diff_sum = 0
        for n1, n2 in zip(nums1, nums2):
            diff = abs(n1 - n2)
            frequency[diff] += 1
            total_diff_sum += diff
            
        # If the total allowable operations can reduce all differences to 0, return 0
        if total_diff_sum <= k:
            return 0
            
        # Process differences from largest to smallest
        for d in range(max_diff, 0, -1):
            if frequency[d] == 0:
                continue
                
            # Count how many items have the current maximum difference `d`
            count = frequency[d]
            
            # If our operations `k` can completely reduce all elements of size `d` to `d - 1`
            if k >= count:
                k -= count
                frequency[d - 1] += count
                frequency[d] = 0
            else:
                # Otherwise, we can only reduce `k` of these elements to `d - 1`
                frequency[d - 1] += k
                frequency[d] -= k
                k = 0
                break # No more operations left
                
        # Calculate the final minimum sum of squared differences
        ans = 0
        for d in range(1, max_diff + 1):
            if frequency[d] > 0:
                ans += frequency[d] * (d * d)
                
        return ans

File: generated_code/cpp/solution_2418_3.cpp
#include <vector>
#include <cmath>
#include <numeric>
#include <algorithm>
#include <map>

using namespace std;

class Solution {
public:
    long long minSumSquareDiff(vector<int>& nums1, vector<int>& nums2, int k1, int k2) {
        int n = nums1.size();
        long long total_k = (long long)k1 + k2;
        
        // Count frequencies of absolute differences
        // Since max value of nums1[i], nums2[i] is 10^5, the max difference is 10^5.
        // A frequency array (or bucket sort approach) yields O(N + MaxDiff) time complexity.
        vector<long long> diff_count(100001, 0);
        long long sum_diff = 0;
        
        for (int i = 0; i < n; ++i) {
            int diff = abs(nums1[i] - nums2[i]);
            diff_count[diff]++;
            sum_diff += diff;
        }
        
        // If the total operations we can perform is greater than or equal to 
        // the sum of all absolute differences, we can reduce all differences to 0.
        if (total_k >= sum_diff) {
            return 0;
        }
        
        // Process differences greedily from largest to smallest
        for (int d = 100000; d > 0; --d) {
            if (diff_count[d] == 0) continue;
            
            long long count = diff_count[d];
            // If we can reduce all elements of size 'd' to size 'd - 1'
            if (total_k >= count) {
                total_k -= count;
                diff_count[d - 1] += count;
                diff_count[d] = 0;
            } else {
                // We can only partially reduce some elements of size 'd' to 'd - 1'
                diff_count[d - 1] += total_k;
                diff_count[d] -= total_k;
                total_k = 0;
                break; // No more operations left
            }
        }
        
        // Calculate the minimum sum of squared differences
        long long ans = 0;
        for (long long d = 1; d <= 100000; ++d) {
            if (diff_count[d] > 0) {
                ans += diff_count[d] * d * d;
            }
        }
        
        return ans;
    }
};